apiserver/internal/handlers/resources/download/download_test.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package download_test
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/internal/handlers/resources/download"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+)
+
+type ValidateSuite struct {
+	testing.IsolationSuite
+
+	fileSystem *MockFileSystem
+}
+
+var _ = gc.Suite(&ValidateSuite{})
+
+func (s *ValidateSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.fileSystem = NewMockFileSystem(ctrl)
+
+	return ctrl
+}
+
+func (s *ValidateSuite) TestValidateResource(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	// Arrange:
+
+	resourceContent := []byte("resource blob content")
+	hasher := sha512.New384()
+	size, err := io.Copy(hasher, strings.NewReader(string(resourceContent)))
+	c.Assert(err, jc.ErrorIsNil)
+	hash := hex.EncodeToString(hasher.Sum(nil))
+
+	downloader := download.NewDownloader(
+		loggertesting.WrapCheckLog(c),
+		s.fileSystem,
+	)
+
+	f, closer := s.expectTmpFile(c)
+	defer closer()
+	s.fileSystem.EXPECT().CreateTemp("", "resource-").Return(f, nil)
+	s.fileSystem.EXPECT().Open(f.Name()).DoAndReturn(func(string) (*os.File, error) {
+		opened, err := os.Open(f.Name())
+		c.Assert(err, jc.ErrorIsNil)
+		return opened, nil
+	})
+	s.fileSystem.EXPECT().Remove(f.Name())
+
+	// Act:
+	reader, err := downloader.Download(
+		context.Background(),
+		io.NopCloser(bytes.NewBuffer(resourceContent)),
+		hash,
+		size,
+	)
+
+	// Assert:
+	c.Assert(err, jc.ErrorIsNil)
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, reader)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(buf.Bytes(), gc.DeepEquals, resourceContent)
+	c.Assert(reader.Close(), jc.ErrorIsNil)
+}
+
+func (s *ValidateSuite) TestGetResourceUnexpectedSize(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	// Arrange:
+
+	resourceContent := []byte("resource blob content")
+	hasher := sha512.New384()
+	_, err := io.Copy(hasher, strings.NewReader(string(resourceContent)))
+	c.Assert(err, jc.ErrorIsNil)
+	hash := hex.EncodeToString(hasher.Sum(nil))
+
+	downloader := download.NewDownloader(
+		loggertesting.WrapCheckLog(c),
+		s.fileSystem,
+	)
+
+	f, closer := s.expectTmpFile(c)
+	defer closer()
```

Comment:
This causes IDE noise as a compiler warning, suggest: `defer func() { _ = Close() }()`

Be explicit when ignoring errors.

Reply:
Makes sense, though in this case the close function returns nothing, so this is correct.

Reply:
doing `defer func() { _ = closer() }()` is the style we recommend because it makes it explicit we are ignoring the error

Reply:
Right, but this particular closer doesn't return an error, it uses an assert in it, so there is no error to check.